File: Java_ChildrenCompetition/src/main/java/service/TestService.java
package service;

import model.*;
import repository.IParticipantRepository;
import repository.ITestRepository;

import java.util.ArrayList;
import java.util.List;


public class TestService implements ITestService<Integer, Test> {
    public final ITestRepository<Integer, Test> testRepository;
    public final IParticipantRepository<Integer, Participant> participantRepository;

    public TestService(ITestRepository<Integer, Test> testRepository, IParticipantRepository<Integer, Participant> participantRepository) {
        this.testRepository = testRepository;
        this.participantRepository = participantRepository;
    }


    @Override
    public List<TestDTO> findAllTestDTOs() {
        Iterable<Test> tests =  testRepository.findAll();

        List<TestDTO> testDTOs = new ArrayList<TestDTO>();

        for(Test test : tests){
            String type = test.getType().getType();
            TestAgeCategory testAgeCategory = test.getCategory();
            String ageCategory = String.valueOf(testAgeCategory.getMinAge()) + " - " + String.valueOf(testAgeCategory.getMaxAge());

            Iterable<Participant> participantsForTest = participantRepository.findAllParticipantsForTest(test.getId());
            int noCompetitors = ((List<Participant>)participantsForTest).size();

            TestDTO testDTO = new TestDTO(type, ageCategory, noCompetitors);
            testDTO.setId(test.getId());
            testDTOs.add(testDTO);
        }

        return testDTOs;
    }
}

File: CSharp_ChildrenCompetitionGUI/CSharp_ChildrenCompetitionGUI/CSharp_ChildrenCompetitionGUI/service/TestService.cs
using System.Collections.Generic;
using CSharp_ChildrenCompetitionGUI.model;
using CSharp_ChildrenCompetitionGUI.repository;

namespace CSharp_ChildrenCompetitionGUI.service
{
    public class TestService : ITestService<int, Test>
    {
        public ITestRepository<int, Test> testRepository;
        public IParticipantRepository<int, Participant> participantRepository;

        public TestService(ITestRepository<int, Test> testRepository, IParticipantRepository<int, Participant> participantRepository)
        {
            this.testRepository = testRepository;
            this.participantRepository = participantRepository;
        }

        public List<TestDTO> findAllTestDTOs()
        {
            // throw new System.NotImplementedException();
            IEnumerable<Test> tests = testRepository.findAll();

            List<TestDTO> testDtos = new List<TestDTO>();

            foreach (Test test in tests)
            {
                string type = test.type.type;
                TestAgeCategory testAgeCategory = test.category;
                string ageCategory = testAgeCategory.minAge.ToString() + " - " + testAgeCategory.maxAge.ToString();

                IEnumerable<Participant> participantsForTest =
                    participantRepository.findAllParticipantsForTest(test.id);

                int noCompetitors = ((List<Participant>) participantsForTest).Count;

                TestDTO testDTO = new TestDTO(type, ageCategory, noCompetitors);
                testDTO.id = test.id;
                testDtos.Add(testDTO);
            }

            return testDtos;
        }
    }
}